lib/galaxy/objectstore/__init__.py

```diff
@@ -577,12 +616,26 @@ def _repr_object_for_exception(self, obj):
         except AttributeError:
             return str(obj)
 
-    def _call_method(self, method, obj, default, default_is_exception,
-            **kwargs):
-        """Check all children object stores for the first one with the dataset."""
-        for key, store in self.backends.items():
-            if store.exists(obj, **kwargs):
-                return store.__getattribute__(method)(obj, **kwargs)
+    def _get_backend(self, obj, **kwargs):
+        """
+        Check all children object stores for the first one with the dataset;
+        it first checks storage media, if given, then evaluates other backends.
+        """
+        for key, backend in self.backends.items():
+            if backend.exists(obj, **kwargs):
+                return backend
+        return None
+
+    def _call_method(self, method, obj, default, default_is_exception, ignore_media=False, **kwargs):
+        if hasattr(obj, "active_storage_media_associations") and \
+                len(obj.active_storage_media_associations) > 0 and \
+                not ignore_media:
+            media = UserObjectStore(obj.active_storage_media_associations, self)
```

Comment:
This may be fine but this piece of code repeated several times seems counter intuitive to me - shouldn't we be hitting the UserObjectStore if we need it and not overriding this stuff at this level in other object stores? How does the UserObjectStore interact with like DiskObjectStores then?

Reply:
I guess you addressed it above with 

> The functionality is leveraging Hierarchical ObjectStore; hence, it is functional only if Hierarchical ObjectStore is configure. However, the hierarchy is applied instance-wide only, and does not affect user’s plugged media configuration;

I don't know... I feel like the abstraction is subtly off here again though.

Reply:
You mean we should not use Hierarchical objectstore? I guess @dannon finds this best,

Reply:
I think a point is to perhaps create a new class, could be a subclass of `HierarchicalObjectStore`, maybe `UserLocalHierarchicalObjectStore`, and have the 'special' behavior happen as a function of the new object?

Reply:
Would the [UserObjectStore class](https://github.com/VJalili/galaxy/blob/6ecc54c56ab0a84b5876c07d11e3c2bd64463f17/lib/galaxy/objectstore/__init__.py#L965) be in the correct direction? The goal of the current implementation is to minimize the footprint, so that a new ObjectStore instance per user is **not** created at higher-up in the call hierarchy (e.g., at job execution level), rather a new instance creation per used is pushed to the ObjectStore level. This way the changes are largely limited to the one method of hierarchical ObjectStore only (i.e., [_call_method](https://github.com/VJalili/galaxy/blob/6ecc54c56ab0a84b5876c07d11e3c2bd64463f17/lib/galaxy/objectstore/__init__.py#L673)).

Reply:
I follow up in https://github.com/galaxyproject/galaxy/pull/9554.